File: Sort/cocktail_sort.py
# Alogrithm for Cocktail Sort to sort element in ascending order

# Cocktail Sort is a variation of Bubble sort. The Bubble sort algorithm always 
# traverses elements from left and moves the largest element to its correct position 
# in first iteration and second largest in second iteration and so on. Cocktail Sort 
# traverses through a given array in both directions alternatively.

def cocktailSort(array):
    arrayLen = len(array)
    check = 1
    while(check == 1):
        check = 0
        for x in range(arrayLen-1):
            if (array[x] > array[x+1]):
                check = 1
                array[x], array[x+1] = array[x+1], array[x]
            if (x == arrayLen - 1):
                break
        if (check == 1):
            check = 0
            for y in reversed(range(arrayLen -1)):
                if array[y] > array[y+1]:
                    check = 1
                    array[y], array[y+1] = array[y+1], array[y]
                if (y == 0):
                    break

numbers = [3,45,1,3,2,8,5,0]
print("Unsorted Numbers: ", numbers)
cocktailSort(numbers)
print("Sorted Numbers: ", numbers)

##|                             |                            |
##| --------------------------- | -------------------------- |
##| Worst case performance      | O(n * n)                   |
##| Best case performance       | O(n )                      |
##| Average case performance    | O(n * n)                   |
##| Space complexity            | O(1)                       |

File: Sort/cocktail_sort.cpp
// Alogrithm for Cocktail Sort to sort element in ascending order

// Cocktail Sort is a variation of Bubble sort. The Bubble sort algorithm always 
// traverses elements from left and moves the largest element to its correct position 
// in first iteration and second largest in second iteration and so on. Cocktail Sort 
// traverses through a given array in both directions alternatively.

#include <iostream>
#include <vector>
using namespace std;

void cocktailSort(vector<int >&v, int n){
    bool sorted = true;

    for(int i=0;i<n-1;i++){
        if(v[i]>v[i+1]){
            swap(v[i],v[i+1]);
            sorted = false;
        }
    }

    if(sorted)
        return;

    sorted = false;

    for(int i=n-1;i>0;i--){
        if(v[i]<v[i-1]){
            swap(v[i],v[i-1]);
            sorted = false;
        }
    }

    if(sorted)
        return;

    cocktailSort(v,n);

}

int main(){
    int n;
    cin>>n;

    if(n<0){
        return 0;
    }

    vector<int > v(n,0);

    for(int i=0;i<n;i++){
        cin>>v[i];
    }

    cocktailSort(v,n);

    for(int i=0;i<n;i++){
        cout<<v[i]<<" ";
    }

    return 0;
}

/*
|                             |                            |
| --------------------------- | -------------------------- |
| Worst case performance      | O(n * n)                 |
| Best case performance       | O(n )                    |
| Average case performance    | O(n * n)                 |
| Space complexity            | O(1)                     |
*/